homeassistant/components/sensor/viaggiatreno.py

```diff
@@ -0,0 +1,190 @@
+        self.hass = hass
+        self._state = None
+        self._attributes = {}
+        self._unit = ''
+        self._icon = ICON
+        self._station_id = station_id
+
+        self.uri = VIAGGIATRENO_ENDPOINT.format(
+            station_id=station_id,
+            train_id=train_id)
+
+        if not name:
+            self._name = "Train {} from {}".format(train_id, station_id)
+        else:
+            self._name = name.format(train_id)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._unit
+
+    @property
+    def device_state_attributes(self):
+        """Return extra attributes."""
+        self._attributes[ATTR_ATTRIBUTION] = CONF_ATTRIBUTION
+        return self._attributes
+
+    @staticmethod
+    def has_departed(data):
+        """Check if the train has actually departed."""
+        try:
+            first_station = data['fermate'][0]
+            if data['oraUltimoRilevamento'] or first_station['effettiva']:
+                return True
+        except ValueError:
+            _LOGGER.error('Cannot fetch first station: %s', data)
+        return False
+
+    @staticmethod
+    def has_arrived(data):
+        """Check if the train has already arrived."""
+        last_station = data["fermate"][-1]
+        if not last_station["effettiva"]:
+            return False
+        return True
+
+    @staticmethod
+    def is_cancelled(data):
+        """Check if the train is cancelled."""
+        if data['tipoTreno'] == 'ST' and data['provvedimento'] == 1:
+            return True
+        return False
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Update state."""
+        uri = self.uri
+        res = yield from async_http_request(self.hass, uri)
+        if res.get('error', ''):
+            if res['error'] == 204:
+                self._state = NO_INFORMATION_STRING
+                self._unit = ''
+            else:
+                self._state = 'Error: {}'.format(res['error'])
+                self._unit = ''
+        else:
+            for i in MONITORED_INFO:
+                self._attributes[i] = res[i]
+
+            if self.is_cancelled(res):
+                self._state = CANCELLED_STRING
+                self._icon = 'mdi:cancel'
+                self._unit = ''
+            elif not self.has_departed(res):
+                self._state = NOT_DEPARTED_STRING
+                self._unit = ''
+            elif self.has_arrived(res):
+                self._state = ARRIVED_STRING
+                self._unit = ''
+            else:
+                self._state = res.get('ritardo', '0')
```

Comment:
What does the `'0'` represent for state? If it's an unknown state, you should set state to `None`.